File: Day #1376/solution.py
# Day 1376: Uniform linked-list shuffle via Fisher-Yates. Time O(n), Space O(n) for the
# index pass (space-over-time variant: O(1) space, O(n^2) by random node selection).
# A deterministic LCG is used so output is reproducible.


class Node:
    def __init__(self, val):
        self.val = val
        self.next = None


_seed = 42


def next_rand():
    global _seed
    _seed = (_seed * 1103515245 + 12345) % 2147483648
    return _seed


def shuffle(head):
    nodes = []
    p = head
    while p:
        nodes.append(p)
        p = p.next
    for i in range(len(nodes) - 1, 0, -1):
        j = next_rand() % (i + 1)
        nodes[i].val, nodes[j].val = nodes[j].val, nodes[i].val
    return head


if __name__ == "__main__":
    head = tail = None
    for v in range(1, 6):
        node = Node(v)
        if head is None:
            head = tail = node
        else:
            tail.next = node
            tail = node
    head = shuffle(head)
    out = []
    p = head
    while p:
        out.append(str(p.val))
        p = p.next
    print(" -> ".join(out))

File: Day #1376/solution.js
// Uniform linked-list shuffle via Fisher-Yates. Time O(n), Space O(n) for the
// index pass (space-over-time variant: O(1) space, O(n^2) by random node selection).
// A deterministic LCG (BigInt for exact 64-bit math) is used so output is reproducible.
class Node {
  constructor(val) { this.val = val; this.next = null; }
}

let seed = 42n;
function nextRand() {
  seed = (seed * 1103515245n + 12345n) % 2147483648n;
  return seed;
}

function shuffle(head) {
  const nodes = [];
  for (let p = head; p; p = p.next) nodes.push(p);
  for (let i = nodes.length - 1; i >= 1; i--) {
    const j = Number(nextRand() % BigInt(i + 1));
    [nodes[i].val, nodes[j].val] = [nodes[j].val, nodes[i].val];
  }
  return head;
}

let head = null, tail = null;
for (let v = 1; v <= 5; v++) {
  const node = new Node(v);
  if (!head) head = tail = node;
  else { tail.next = node; tail = node; }
}
head = shuffle(head);
const out = [];
for (let p = head; p; p = p.next) out.push(p.val);
console.log(out.join(" -> "));
